internal/contentmanager/source/dynamicsource.go

```diff
@@ -0,0 +1,223 @@
+	Predicates []predicate.Predicate
+	// OnPostSyncError is the callback function that is used when the source
+	// initially synced successfully and later encountered an error
+	OnPostSyncError func(context.Context)
+}
+
+func NewDynamicSource(cfg DynamicSourceConfig) *dynamicInformerSource {
+	return &dynamicInformerSource{
+		cfg:        cfg,
+		errChan:    make(chan error, 1),
+		syncedChan: make(chan struct{}, 1),
+	}
+}
+
+// dynamicInformerSource is an implementation of the
+// ReaderCloserSyncingSource interface. It is used
+// to create an informer, using the provided dynamic informer
+// factory, for the configured GroupVersionResource.
+//
+// The informer is configured with a WatchEventErrorHandler that
+// stops the informer, and if it had previously synced successfully
+// it attempts to requeue the provided Owner for reconciliation and
+// calls the provided OnWatchError function.
+type dynamicInformerSource struct {
+	cfg        DynamicSourceConfig
+	errOnce    sync.Once
+	errChan    chan error
+	err        error
+	syncedChan chan struct{}
+	mu         sync.Mutex
+	synced     bool
+	cancel     context.CancelFunc
+	ctx        context.Context
+	informer   informers.GenericInformer
+	started    bool
+}
+
+func (dis *dynamicInformerSource) String() string {
+	dis.mu.Lock()
+	defer dis.mu.Unlock()
+	return fmt.Sprintf("contentmanager.DynamicInformerSource - GVR: %s, Owner Type: %T, Owner Name: %s", dis.cfg.GVR, dis.cfg.Owner, dis.cfg.Owner.GetName())
+}
+
+func (dis *dynamicInformerSource) Start(ctx context.Context, q workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+	// Lock to check if this source has already been started.
+	// This should be the only time we need to lock, even if this
+	// function is called concurrently. This is because the first
+	// call that is able to lock and set the started value to
+	// true would "win" and be the one that continues on.
+	dis.mu.Lock()
+	if dis.started {
+		dis.mu.Unlock()
+		return errors.New("already started")
+	}
+
+	dis.started = true
+	dis.mu.Unlock()
+
+	informerCtx, cancel := context.WithCancel(ctx)
+
+	gInf := dis.cfg.DynamicInformerFactory.ForResource(dis.cfg.GVR)
+	dis.errChan = make(chan error, 1)
+	dis.syncedChan = make(chan struct{}, 1)
+	dis.informer = gInf
+	dis.cancel = cancel
+	dis.ctx = informerCtx
+	eventHandler := source.NewEventHandler(ctx, q, dis.cfg.Handler, dis.cfg.Predicates)
+
+	// If we encounter an error during the watch we will:
+	// - Capture the error
+	// - cancel the informer
+	// requeuing of the ClusterExtension should happen by the
+	// WaitForSync function returning an error
+	// Only if we have successfully synced in the past should we
+	// requeue the ClusterExtension
+	sharedIndexInf := gInf.Informer()
+	err := sharedIndexInf.SetWatchErrorHandler(func(r *cgocache.Reflector, err error) {
+		dis.mu.Lock()
+		dis.errOnce.Do(func() {
+			dis.errChan <- err
+			dis.err = err
+			close(dis.errChan)
+		})
+		dis.cancel()
```

Comment:
move this below the call to the OnPostSyncError so that we don't pass it a cancelled context.

Reply:
addressed in https://github.com/operator-framework/operator-controller/pull/1119/commits/10eaf1f8ce5394edf8392f29f2e21710bf38b8fc